{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importing and initializing\n",
    "import networkx as nx\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy \n",
    "from node2vec import Node2Vec\n",
    "import umap\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib.cm as cm\n",
    "from gensim.models import Word2Vec\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/6q/0qnxsny943385fwzz738241c0000gn/T/ipykernel_86843/547375681.py:3: DtypeWarning: Columns (8) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  edge_df = pd.read_csv(\"2024_04_24_meta_edge_sourcetarget_disease_pub_pli_reciprocal.csv\")\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Stored 'geneGraph' (DiGraph)\n",
      "Stored 'node_df' (DataFrame)\n",
      "Stored 'edge_df' (DataFrame)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/anusha/Desktop/Neuroinformatics_Internship/HopkinsGeneClassificationProj/env/lib/python3.12/site-packages/IPython/extensions/storemagic.py:229: UserWarning: This is now an optional IPython functionality, setting autorestore/geneGraph requires you to install the `pickleshare` library.\n",
      "  db[ 'autorestore/' + arg ] = obj\n",
      "/Users/anusha/Desktop/Neuroinformatics_Internship/HopkinsGeneClassificationProj/env/lib/python3.12/site-packages/IPython/extensions/storemagic.py:229: UserWarning: This is now an optional IPython functionality, setting autorestore/node_df requires you to install the `pickleshare` library.\n",
      "  db[ 'autorestore/' + arg ] = obj\n",
      "/Users/anusha/Desktop/Neuroinformatics_Internship/HopkinsGeneClassificationProj/env/lib/python3.12/site-packages/IPython/extensions/storemagic.py:229: UserWarning: This is now an optional IPython functionality, setting autorestore/edge_df requires you to install the `pickleshare` library.\n",
      "  db[ 'autorestore/' + arg ] = obj\n"
     ]
    }
   ],
   "source": [
    "# Locating, initializing, and storing data and graphs\n",
    "node_df = pd.read_csv(\"2024_05_05_meta_node_data.csv\")\n",
    "edge_df = pd.read_csv(\"2024_04_24_meta_edge_sourcetarget_disease_pub_pli_reciprocal.csv\")\n",
    "geneGraph = nx.DiGraph()\n",
    "%store geneGraph\n",
    "%store node_df\n",
    "%store edge_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>disease_assoc_cat</th>\n",
       "      <th>publication_count</th>\n",
       "      <th>ortholog_count</th>\n",
       "      <th>all_ortho_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7414</th>\n",
       "      <td>VCL</td>\n",
       "      <td>disease_assoc</td>\n",
       "      <td>279</td>\n",
       "      <td>5</td>\n",
       "      <td>{'hsapien': {'entrezid': 7414, 'symbol': 'VCL'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4626</th>\n",
       "      <td>MYH8</td>\n",
       "      <td>disease_assoc</td>\n",
       "      <td>36</td>\n",
       "      <td>4</td>\n",
       "      <td>{'hsapien': {'entrezid': 4626, 'symbol': 'MYH8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9722</th>\n",
       "      <td>NOS1AP</td>\n",
       "      <td>disease_assoc</td>\n",
       "      <td>141</td>\n",
       "      <td>5</td>\n",
       "      <td>{'hsapien': {'entrezid': 9722, 'symbol': 'NOS1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9891</th>\n",
       "      <td>NUAK1</td>\n",
       "      <td>non_disease_assoc</td>\n",
       "      <td>70</td>\n",
       "      <td>5</td>\n",
       "      <td>{'hsapien': {'entrezid': 9891, 'symbol': 'NUAK...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81788</th>\n",
       "      <td>NUAK2</td>\n",
       "      <td>disease_assoc</td>\n",
       "      <td>46</td>\n",
       "      <td>2</td>\n",
       "      <td>{'hsapien': {'entrezid': 81788, 'symbol': 'NUA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       symbol  disease_assoc_cat  publication_count  ortholog_count  \\\n",
       "id                                                                    \n",
       "7414      VCL      disease_assoc                279               5   \n",
       "4626     MYH8      disease_assoc                 36               4   \n",
       "9722   NOS1AP      disease_assoc                141               5   \n",
       "9891    NUAK1  non_disease_assoc                 70               5   \n",
       "81788   NUAK2      disease_assoc                 46               2   \n",
       "\n",
       "                                         all_ortho_count  \n",
       "id                                                        \n",
       "7414   {'hsapien': {'entrezid': 7414, 'symbol': 'VCL'...  \n",
       "4626   {'hsapien': {'entrezid': 4626, 'symbol': 'MYH8...  \n",
       "9722   {'hsapien': {'entrezid': 9722, 'symbol': 'NOS1...  \n",
       "9891   {'hsapien': {'entrezid': 9891, 'symbol': 'NUAK...  \n",
       "81788  {'hsapien': {'entrezid': 81788, 'symbol': 'NUA...  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "node_df.set_index('id',inplace=True)\n",
    "node_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Finding a duplicate edges\n",
    "\n",
    "# edges = set()\n",
    "# for index, row in edge_df.iterrows():\n",
    "#     gene = row['source_target_key']\n",
    "#     edges.add(gene)\n",
    "\n",
    "# print(len(edges))\n",
    "\n",
    "# # There are 430073 duplicate edges."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Constructing a  graph with first edges\n",
    "# This method would only work if you take the nodes in order because the index of edge_list and edge_df (and first_edges) must match\n",
    "# Otherwise, the for loop would not work properly and the arrows would be wrong\n",
    "\n",
    "# 1048576\n",
    "\n",
    "# Take the first edges -- this is a list\n",
    "first_edges = list(edge_df.iloc[:1000]['source_target_key'])\n",
    "\n",
    "# Convert each edge into a tuple\n",
    "edge_tuples = lambda x: tuple(map(int,x.split(';')))\n",
    "\n",
    "# Put all those tuples in a list\n",
    "edge_list = list(map(edge_tuples, first_edges))\n",
    "\n",
    "# Add edges to graph\n",
    "geneGraph.add_edges_from(edge_list)\n",
    "\n",
    "# Weighting edges on DIOPT scores\n",
    "diopt_dict = {}\n",
    "\n",
    "for index, edge in enumerate(edge_list):\n",
    "    diopt_dict[edge] = edge_df.iloc[index]['interaction_count']\n",
    "\n",
    "nx.set_edge_attributes(geneGraph, diopt_dict, 'weight')\n",
    "\n",
    "# Add bidirectional edges based on reciprocal value\n",
    "bidirectional_edges = []\n",
    "\n",
    "for index, edge in enumerate(edge_list):\n",
    "    reciprocal = edge_df.iloc[index]['reciprocal']\n",
    "if reciprocal:\n",
    "    reversed_edge = (edge[1], edge[0])\n",
    "    bidirectional_edges.append(reversed_edge)\n",
    "\n",
    "# Adding edges to graph\n",
    "geneGraph.add_edges_from(bidirectional_edges)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nodes without edges\n",
    "all_node_ids = set(node_df.index)\n",
    "nodes_with_edges = set(edge_df['source']).union(set(edge_df['target']))\n",
    "\n",
    "missing_nodes = all_node_ids - nodes_with_edges\n",
    "print(f'These nodes do not have edges: {missing_nodes}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEGREE DISTRIBUTION\n",
    "\n",
    "# Degrees of each node\n",
    "node_degrees = geneGraph.degree(geneGraph.nodes)\n",
    "\n",
    "# Finding degree distribution\n",
    "degrees = []\n",
    "for node, degree in node_degrees:\n",
    "    degrees.append(degree)\n",
    "\n",
    "# The frequency of each degree value\n",
    "degree_freq = {}\n",
    "for degree in degrees:\n",
    "    if degree in degree_freq:\n",
    "        degree_freq[degree] += 1\n",
    "    else:\n",
    "        degree_freq[degree] = 1\n",
    "\n",
    "sorted_degree_freq = dict(sorted(degree_freq.items(), key=lambda item: item[1], reverse=True))\n",
    "\n",
    "# Convert to dictionary, of nodes:degrees\n",
    "node_degree_dict = {}\n",
    "for key, value in node_degrees:\n",
    "    if key not in node_degree_dict:\n",
    "        node_degree_dict[key] = value\n",
    "\n",
    "            \n",
    "    # for key, value in sorted_degree_freq.items():\n",
    "    #     print(f\"{value} instances of {key} degrees.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEGREE DISTRIBUTION - bar chart\n",
    "y_frequencies = []\n",
    "x_degrees = []\n",
    "for degree, frequency in degree_freq.items():\n",
    "    y_frequencies.append(frequency)\n",
    "    x_degrees.append(degree)\n",
    "# Sorting the degrees to ensure proper ordering\n",
    "x_degrees, y_frequencies = zip(*sorted(zip(x_degrees, y_frequencies)))\n",
    "\n",
    "# Limiting to the first 100 items\n",
    "x_degrees = x_degrees[10:]\n",
    "y_frequencies = y_frequencies[10:]\n",
    "plt.figure(figsize=(12,8))\n",
    "plt.bar(x_degrees, y_frequencies)\n",
    "plt.xticks()\n",
    "plt.xlabel('Degree of node')\n",
    "plt.ylabel('Frequencies')\n",
    "plt.title('Node Degree Distribution (above 10 degrees)')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# DEGREE DISTRIBUTION - histogram\n",
    "plt.figure(figsize=(12,8))\n",
    "\n",
    "plt.hist(degrees, bins = 10, color = 'blue')\n",
    "plt.xlabel('Degrees')\n",
    "plt.ylabel('Frequencies')\n",
    "plt.title('Node Degree Distribution')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Nodes color spectrum based on publication count\n",
    "\n",
    "# Map values to colors based on color map\n",
    "def get_colors(nodes, cmap):\n",
    "    if not nodes:\n",
    "        return {}\n",
    "    values = list(nodes.values())\n",
    "    norm = mcolors.Normalize(vmin=min(numpy.log1p(values)), vmax=max(numpy.log1p(values)))   # normalizing all values on a scale of 0 to 1 for coloring\n",
    "    cmap = cm.get_cmap(cmap)\n",
    "    return {node: cmap(norm(value)) for node, value in nodes.items()}\n",
    "\n",
    "# Apply colormap to each category's nodes and collect colors\n",
    "def nodes_apply_colormap(G):\n",
    "    # Creating a dict and storing all genes\n",
    "    disease_assoc = {}\n",
    "    non_disease_assoc = {}\n",
    "    non_omim = {}\n",
    "    merged = {}\n",
    "    for node in G.nodes:\n",
    "        if node not in node_df.index:\n",
    "            merged[node] = 1\n",
    "            continue\n",
    "        disease_cat = node_df.loc[node]['disease_assoc_cat']\n",
    "        if disease_cat == 'disease_assoc':\n",
    "            disease_assoc[node] = node_df.loc[node]['publication_count']\n",
    "        elif disease_cat == 'non_disease_assoc':\n",
    "            non_disease_assoc[node] = node_df.loc[node]['publication_count']\n",
    "        elif disease_cat == 'non_omim':\n",
    "            non_omim[node] = node_df.loc[node]['publication_count']\n",
    "        else:\n",
    "            merged[node] = 1\n",
    "    # Identifying base colors for each category\n",
    "    base_colors = {'disease_assoc':'Reds', 'non_disease_assoc':'Blues', 'non_omim': 'Grays', 1:'Grays'}\n",
    "\n",
    "    # Mapping the categories to each other to normalize\n",
    "    all_nodes = {'disease_assoc': disease_assoc,'non_disease_assoc': non_disease_assoc, 'non_omim': non_omim, 1:merged}\n",
    "\n",
    "    # Storing node colors in a list\n",
    "    node_colors = []\n",
    "    for category, nodes in all_nodes.items():\n",
    "        if category not in base_colors:\n",
    "            print(f\"Category {category} is not in base colors\")\n",
    "        else:\n",
    "            cmap_name = base_colors[category]\n",
    "            colors = get_colors(nodes, cmap_name)\n",
    "            if category == 'unknown':\n",
    "                colors = {node: (0, 0, 0, 1) for node in nodes} # Making 'unknown' nodes black\n",
    "            node_colors.extend(colors[node] for node in G.nodes if node in colors)\n",
    "    return node_colors\n",
    "\n",
    "node_colors = nodes_apply_colormap(geneGraph)\n",
    "print(node_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'gray', 'tab:red', 'tab:red', 'tab:red', 'gray', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'gray', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'gray', 'tab:blue', 'gray', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'gray', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'gray', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'gray', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'gray', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'gray', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'gray', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'gray', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'gray', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'gray', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'tab:red', 'tab:red', 'gray', 'tab:red', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:red', 'gray', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'tab:blue', 'gray', 'tab:blue', 'tab:red']\n"
     ]
    }
   ],
   "source": [
    "# Creating a dictionary that assigns nodes to their disease association in a dict\n",
    "nodes_list = list(geneGraph.nodes)\n",
    "disease_cat_dict = {}\n",
    "for node in nodes_list:\n",
    "    disease_assoc_cat = node_df.loc[node]['disease_assoc_cat']\n",
    "    if disease_assoc_cat == 'disease_assoc':\n",
    "        disease_cat_dict[node] = {'category':'disease_assoc'}\n",
    "    elif disease_assoc_cat == 'non_disease_assoc':\n",
    "        disease_cat_dict[node] = {'category':'non_disease_assoc'}\n",
    "    elif disease_assoc_cat == 'non_omim':\n",
    "        disease_cat_dict[node] = {'category':'non_omim'}\n",
    "    else:\n",
    "        print ('failed')\n",
    "\n",
    "# Assigning the attrubutes to the nodes\n",
    "nx.set_node_attributes(geneGraph, disease_cat_dict)\n",
    "\n",
    "# Creating a color map for the categories to colors\n",
    "node_color_map = {'disease_assoc':'tab:red','non_disease_assoc':'tab:blue','non_omim':'gray'}\n",
    "node_colors = [node_color_map[geneGraph.nodes[node]['category']] for node in geneGraph.nodes]\n",
    "\n",
    "print(node_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Coloring edges by 6 sections of interaction_type\n",
    "\n",
    "# GPT's 6 sections of the 84 items\n",
    "genetic_interactions = ['cisphenotypic genetic ','mutual genetic over-su','association', 'genetic interaction', 'genetic suppression', 'genetic suppression (p', 'genetic suppression (c', 'genetic enhancement', 'genetic epistasis (sen', 'phenotypic suppression', 'asynthetic genetic int', 'cisphenotypic co-suppr', 'opposing genetic epist', 'minimal genetic epista', 'synthetic growth defec', 'transphenotypic enhanc', 'transphenotypic geneti', 'genetic over-suppressi', 'negative genetic inter', 'positive genetic inter']\n",
    "physical_interactions = ['self interaction','physical association', 'direct interaction', 'colocalization', 'proximity', 'protein cleavage', 'disulfide bond', 'Co-localization', 'Co-fractionation', 'Co-purification', 'Co-crystal Structure', 'Protein-peptide', 'Protein-RNA']\n",
    "post_translational_modifications = ['phosphorylation','demethylation reaction', 'phosphorylation reacti', 'acetylation reaction', 'covalent binding', 'cleavage reaction', 'dephosphorylation reac', 'ubiquitination reactio', 'methylation reaction', 'deacetylation', 'cleavage', 'dephosphorylation', 'methylation', 'ubiquitination', 'deformylation reaction']\n",
    "experimental_methods = ['Affinity Capture-RNA', 'Affinity Capture-Weste', 'Affinity Capture-MS', 'Two-hybrid', 'Biochemical Activity', 'Reconstituted Complex', 'Negative Genetic', 'Positive Genetic', 'PCA', 'FRET', 'Proximity Label-MS', 'Far Western', 'Affinity Capture-Lumin']\n",
    "synthetic_rescue_and_dosage_interactions = ['dosage lethality (sens','synthetic genetic inte', 'mutual genetic enhance', 'maximal genetic epista', 'synthetic rescue (sens', 'dosage rescue (sensu B', 'synthetic rescue (sens', 'synthetic lethality (s', 'synthetic haploinsuffi', 'dosage rescue (sensu b', 'dosage growth defect (', 'Synthetic Rescue', 'Synthetic Lethality', 'Dosage Lethality', 'Dosage Rescue',  'Dosage Growth Defect', 'Synthetic Growth Defec', 'Synthetic Haploinsuffi']\n",
    "phenotypic_interactions = ['monophenotypic genetic', 'phenotypic enhancement', 'putative self interact', 'cisphenotypic inter-su', 'cisphenotypic inter-suppr', 'Phenotypic Suppression', 'Phenotypic Enhancement']\n",
    "\n",
    "def color_edge_section(G):\n",
    "    # Creating a dict that assigns each interaction_type to its respective section\n",
    "    interaction_section_dict = {}\n",
    "    for index, row in edge_df.iterrows():\n",
    "        edge = (row['source'], row['target'])\n",
    "        interaction_type = row['interaction_type']\n",
    "        if interaction_type in genetic_interactions:\n",
    "            interaction_section_dict[edge] = {'category':'Genetic Interaction'}\n",
    "        elif interaction_type in physical_interactions:\n",
    "            interaction_section_dict[edge] = {'category':'Physical Interaction'}\n",
    "        elif interaction_type in post_translational_modifications:\n",
    "            interaction_section_dict[edge] = {'category':'Post Translational Modification'}\n",
    "        elif interaction_type in experimental_methods:\n",
    "            interaction_section_dict[edge] = {'category':'Experimental Method'}\n",
    "        elif interaction_type in synthetic_rescue_and_dosage_interactions:\n",
    "            interaction_section_dict[edge] = {'category':'Synthetic Rescue and Dosage Interaction'}\n",
    "        elif interaction_type in phenotypic_interactions:\n",
    "            interaction_section_dict[edge] = {'category':'Phenotypic Interaction'}\n",
    "        else:\n",
    "            print(f'edge color failed for {interaction_type}')\n",
    "            interaction_section_dict[edge] = {'category':'Unknown'}\n",
    "\n",
    "\n",
    "    # Assigning attributes to edge\n",
    "    nx.set_edge_attributes(G, interaction_section_dict)\n",
    "\n",
    "    # Creating a color map that maps each section to a color\n",
    "    edge_section_color_map = {'Genetic Interaction':'orange', 'Physical Interaction':'yellow', 'Post Translational Modification':'green','Experimental Method':'purple','Synthetic Rescue and Dosage Interaction':'pink','Phenotypic Interaction':'brown', 'Unknown':'black'}\n",
    "\n",
    "    edge_colors_int_section = [edge_section_color_map.get(G.edges[edge].get('category', 'Unknown'), 'black') for edge in G.edges]    \n",
    "    # edge_colors_int_section = [edge_section_color_map[G.edges[edge]['category']] for edge in G.edges]\n",
    "    return edge_colors_int_section\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Coloring  edges by 5 top interaction types\n",
    "\n",
    "# interaction_type_dict = {}\n",
    "# for index, edge in enumerate(geneGraph.edges):\n",
    "#     interaction_type = edge_df.loc[index]['interaction_type']\n",
    "#     if interaction_type == 'Negative Genetic':\n",
    "#         interaction_type_dict[edge] = {'category':'Negative Genetic'}\n",
    "#     elif interaction_type == 'physical association':\n",
    "#         interaction_type_dict[edge] = {'category':'physical association'}\n",
    "#     elif interaction_type == 'Affinity Capture-MS':\n",
    "#         interaction_type_dict[edge] = {'category':'Affinity Capture-MS'}\n",
    "#     elif interaction_type == 'direct interaction':\n",
    "#         interaction_type_dict[edge] = {'category':'direct interaction'}\n",
    "#     elif interaction_type == 'Positive Genetic':\n",
    "#         interaction_type_dict[edge] = {'category':'Positive Genetic'}\n",
    "#     else:\n",
    "#         interaction_type_dict[edge] = {'category':'other'}\n",
    "\n",
    "# # Assigning attributes to edge\n",
    "# nx.set_edge_attributes(geneGraph, interaction_type_dict)\n",
    "\n",
    "# # Creating a color map to map the interaction types to colors\n",
    "# edge_type_color_map = {'Negative Genetic':'orange','physical association':'yellow','Affinity Capture-MS':'green','direct interaction':'purple', 'Positive Genetic':'pink', 'other':'black'}\n",
    "# edge_colors_int_type = [edge_type_color_map[geneGraph.edges[edge]['category']]for edge in geneGraph.edges]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Coloring edges by interaction_cat\n",
    "\n",
    "# # Creating a dictionary that assigns edges to their interaction_cat in a dict\n",
    "# interaction_cat_dict = {}\n",
    "# for index,edge in enumerate(geneGraph.edges):\n",
    "#     interaction_cat = edge_df.loc[index]['interaction_cat']\n",
    "#     if interaction_cat == 'physical':\n",
    "#         interaction_cat_dict[edge] = {'category':'physical'}\n",
    "#     elif interaction_cat == 'genetic':\n",
    "#         interaction_cat_dict[edge] = {'category':'genetic'}\n",
    "#     else:\n",
    "#         print('failed')\n",
    "\n",
    "# # Assigning attributes to edge\n",
    "# nx.set_edge_attributes(geneGraph, interaction_cat_dict)\n",
    "\n",
    "# # Creating a color map for the categories to colors\n",
    "# edge_cat_color_map = {'physical':'orange', 'genetic':'green'}\n",
    "# edge_colors_int_cat = [edge_cat_color_map[geneGraph.edges[edge]['category']] for edge in geneGraph.edges]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Thickness of the edges correspond to interaction_count\n",
    "\n",
    "# def edge_thickness(G):\n",
    "#     # Creating a dictionary that assigns edges to their interaction_count in a dict\n",
    "#     interaction_count_dict = {}\n",
    "#     for index, edge in enumerate(G.edges):\n",
    "#         interaction_count = edge_df.iloc[index]['interaction_count']\n",
    "#         if interaction_count == 1:\n",
    "#             interaction_count_dict[edge] = {'count':1}\n",
    "#         elif interaction_count == 2:\n",
    "#             interaction_count_dict[edge] = {'count':2}\n",
    "#         elif interaction_count == 3:\n",
    "#             interaction_count_dict[edge] = {'count':3}\n",
    "#         elif interaction_count == 4:\n",
    "#             interaction_count_dict[edge] = {'count':4}\n",
    "#         elif interaction_count == 5:\n",
    "#             interaction_count_dict[edge] = {'count':5}\n",
    "#         elif interaction_count == 6:\n",
    "#             interaction_count_dict[edge] = {'count':6}\n",
    "#         else:\n",
    "#             print('failed')\n",
    "\n",
    "#     # Assigning attributes to edge\n",
    "#     nx.set_edge_attributes(G, interaction_count_dict)\n",
    "\n",
    "#     # Creating a width map for the categories to colors\n",
    "#     # edge_width_map = {1:0.02,2:0.07,3:.12,4:.17,5:.22,6:.27}    #thinner\n",
    "#     edge_width_map = {1:0.2,2:0.7,3:1.2,4:1.7,5:2.2,6:2.7}    #thicker\n",
    "#     edge_widths = [edge_width_map[G.edges[edge]['count']] for edge in G.edges]\n",
    "#     return edge_widths\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Changing style of edges by interaction_cat\n",
    "\n",
    "def edge_style(G):\n",
    "    # Creating a dictionary that assigns edges to their interaction_cat in a dict\n",
    "    int_cat_dict = {}\n",
    "    for index,edge in enumerate(G.edges):\n",
    "        interaction_cat = edge_df.loc[index]['interaction_cat']\n",
    "        if interaction_cat == 'physical':\n",
    "            int_cat_dict[edge] = {'category':'physical'}\n",
    "        elif interaction_cat == 'genetic':\n",
    "            int_cat_dict[edge] = {'category':'genetic'}\n",
    "        else:\n",
    "            print('category not found')\n",
    "            int_cat_dict[edge] = {'category':'unknown'}\n",
    "\n",
    "    # Assigning attributes to edge\n",
    "    nx.set_edge_attributes(G, int_cat_dict)\n",
    "\n",
    "    # Creating a style map for the categories to colors\n",
    "    edge_cat_style_map = {'physical':'dotted', 'genetic':'solid', 'unknown':'dashed'}\n",
    "    edge_styles_int_cat = [edge_cat_style_map[G.edges[edge]['category']] for edge in G.edges]\n",
    "    return edge_styles_int_cat\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edge_colors_int_section = color_edge_section(geneGraph)\n",
    "edge_styles_int_cat = edge_style(geneGraph)\n",
    "node_colors = nodes_apply_colormap(geneGraph)\n",
    "\n",
    "print(len(edge_colors_int_section))\n",
    "print(len(edge_styles_int_cat))\n",
    "print(len(geneGraph.edges))\n",
    "print(len(edge_list))\n",
    "\n",
    "print(len(geneGraph.nodes))\n",
    "print(len(node_colors))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Spring layout - colored by interaction_cat\n",
    "edge_colors_int_section = color_edge_section(geneGraph)\n",
    "edge_styles_int_cat = edge_style(geneGraph)\n",
    "node_colors = nodes_apply_colormap(geneGraph)\n",
    "pos = nx.spring_layout(geneGraph, k=0.7, iterations=15, seed = 400) #scale\n",
    "fig, ax = plt.subplots(figsize=(12, 8))\n",
    "nx.draw(geneGraph, pos, with_labels=False, width = 0.1, node_color=node_colors, edge_color = edge_colors_int_section, style = edge_styles_int_cat, node_size=5) #, font_size = 10, font_color='black')\n",
    "\n",
    "# # Showing the color bars\n",
    "# for category, nodes in all_nodes.items():\n",
    "#     cmap_name = base_colors[category]\n",
    "#     values = list(nodes.values())\n",
    "#     norm = mcolors.Normalize(vmin=min(values), vmax=max(values))\n",
    "#     cmap = cm.get_cmap(cmap_name)\n",
    "#     sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "#     sm.set_array([])\n",
    "#     plt.colorbar(sm, ax=ax, label=category)\n",
    "\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #Spring layout - colored by 5 top interaction types\n",
    "# node_colors = [node_color_map[geneGraph.nodes[node]['category']] for node in geneGraph.nodes]\n",
    "# geneGraph.add_edges_from(edge_list)\n",
    "# pos = nx.spring_layout(geneGraph, k=0.7, iterations=15, seed = 400) #scale\n",
    "# nx.draw(geneGraph, pos, with_labels=False, width = edge_widths, edge_color = edge_colors_int_type, node_color=node_colors, node_size=5) #, font_size = 10, font_color='black')\n",
    "# plt.figure(figsize=(100, 50))\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Spring layout - colored by 6 interaction type sections\n",
    "# node_colors = [node_color_map[geneGraph.nodes[node]['category']] for node in geneGraph.nodes]\n",
    "# geneGraph.add_edges_from(edge_list)\n",
    "# pos = nx.spring_layout(geneGraph, k=0.7, iterations=15, seed = 400) #scale\n",
    "# nx.draw(geneGraph, pos, with_labels=False, width = edge_widths, edge_color = edge_colors_int_section, node_color=node_colors, node_size=5) #, font_size = 10, font_color='black')\n",
    "# plt.figure(figsize=(100, 50))\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Spiral (equidistant)\n",
    "# pos = nx.spiral_layout(geneGraph, scale=1, center=None, dim=2, resolution=0.35, equidistant=True)\n",
    "# nx.draw(geneGraph, pos, with_labels=False, width = edge_widths, edge_color = edge_colors_int_cat, node_color=node_colors, node_size=5) #, font_size = 10, font_color='black')\n",
    "# plt.figure(figsize=(100, 50))\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Spiral (sparkler)\n",
    "# node_colors = [node_color_map[geneGraph.nodes[node]['category']] for node in geneGraph.nodes]\n",
    "# geneGraph.add_edges_from(edge_list)\n",
    "# pos = nx.spiral_layout(geneGraph, scale=1, center=None, dim=2, resolution=0.35, equidistant=False)\n",
    "# nx.draw(geneGraph, pos, with_labels=False, width = edge_widths, edge_color = edge_colors_int_cat, node_color=node_colors, node_size=20) #, font_size = 10, font_color='black')\n",
    "# plt.figure(figsize=(100, 50))\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Circular\n",
    "# node_colors = [node_color_map[geneGraph.nodes[node]['category']] for node in geneGraph.nodes]\n",
    "# geneGraph.add_edges_from(edge_list)\n",
    "# pos = nx.circular_layout(geneGraph, scale=1, center=None, dim=2)\n",
    "# nx.draw(geneGraph, pos, with_labels=False, width = edge_widths, edge_color = edge_colors_int_cat, node_color=node_colors, node_size=20) #, font_size = 10, font_color='black')\n",
    "# plt.figure(figsize=(100, 50))\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Kamada Kawai\n",
    "# node_colors = [node_color_map[geneGraph.nodes[node]['category']] for node in geneGraph.nodes]\n",
    "# geneGraph.add_edges_from(edge_list)\n",
    "# pos = nx.kamada_kawai_layout(geneGraph, dist=None, scale=1, center=None, dim=2)\n",
    "# nx.draw(geneGraph, pos, with_labels=False, width = edge_widths, edge_color = edge_colors_int_cat, node_color=node_colors, node_size=5) #, font_size = 10, font_color='black')\n",
    "# plt.figure(figsize=(100, 50))\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Multipartite (color grouping)\n",
    "# node_colors = [node_color_map[geneGraph.nodes[node]['category']] for node in geneGraph.nodes]\n",
    "# geneGraph.add_edges_from(edge_list)\n",
    "# pos = nx.multipartite_layout(geneGraph, subset_key='category', align='vertical', scale=1, center=None)\n",
    "# nx.draw(geneGraph, pos, with_labels=False, width = edge_widths, edge_color = edge_colors_int_cat, node_color=node_colors, node_size=15) #, font_size = 10, font_color='black')\n",
    "# plt.figure(figsize=(100, 50))\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Spectral clustering\n",
    "# node_colors = [node_color_map[geneGraph.nodes[node]['category']] for node in geneGraph.nodes]\n",
    "# geneGraph.add_edges_from(edge_list)\n",
    "# pos = nx.spectral_layout(geneGraph, scale=1, center=None, dim=2)\n",
    "# nx.draw(geneGraph, pos, with_labels=False, width = edge_widths, edge_color = edge_colors_int_cat, node_color=node_colors, node_size=70) #, font_size = 10, font_color='black')\n",
    "# plt.figure(figsize=(100, 50))\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SPARSIFICATION - dedensify\n",
    "geneGraphSparse, sparse_nodes = nx.dedensify(geneGraph, threshold=2)\n",
    "print('Original graph nodes:',len(geneGraph.nodes))\n",
    "print('Sparse graph nodes:',len(geneGraphSparse.nodes))\n",
    "print('\\nOriginal graph edges:', len(geneGraph.edges))\n",
    "print('Sparse graph edges:',len(geneGraphSparse.edges))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Original graph\n",
    "\n",
    "pos = nx.spring_layout(geneGraph, k=0.7, iterations=15, seed = 400) #scale\n",
    "fig, ax = plt.subplots(figsize=(12, 8))\n",
    "nx.draw(geneGraph, pos, with_labels=False, width = 0.1, node_color=node_colors, edge_color = edge_colors_int_section, style = edge_styles_int_cat, node_size=8) #, font_size = 10, font_color='black')\n",
    "\n",
    "# Showing the color bars\n",
    "for category, nodes in all_nodes.items():\n",
    "    cmap_name = base_colors[category]\n",
    "    values = list(nodes.values())\n",
    "    norm = mcolors.Normalize(vmin=min(values), vmax=max(values))\n",
    "    cmap = cm.get_cmap(cmap_name)\n",
    "    sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "    sm.set_array([])\n",
    "    plt.colorbar(sm, ax=ax, label=category)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sparse_edge_colors = color_edge_section(geneGraphSparse)\n",
    "sparse_edge_styles = edge_style(geneGraphSparse)\n",
    "sparse_node_colors = nodes_apply_colormap(geneGraphSparse)\n",
    "\n",
    "print(len(sparse_node_colors))\n",
    "print(len(sparse_edge_colors))\n",
    "print(len(sparse_edge_styles))\n",
    "\n",
    "print(len(geneGraphSparse.edges))\n",
    "print(len(geneGraphSparse.nodes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Finding nodes that are the same in geneGraphSparse and new nodes that were created from merging\n",
    "# consistent_nodes = []\n",
    "# merged_nodes = []\n",
    "# for node in geneGraphSparse.nodes():\n",
    "#     if node in geneGraph.nodes():\n",
    "#         consistent_nodes.append(node)\n",
    "#     elif node not in geneGraph.nodes():\n",
    "#         merged_nodes.append(node)\n",
    "\n",
    "# print(len(consistent_nodes))\n",
    "# print(len(merged_nodes))\n",
    "\n",
    "# consistent_edges = []\n",
    "# merged_edges = []\n",
    "# for edge in geneGraphSparse.edges():\n",
    "#     if edge in geneGraph.edges():\n",
    "#         consistent_edges.append(edge)\n",
    "#     elif edge not in geneGraph.edges():\n",
    "#         merged_edges.append(edge)\n",
    "\n",
    "# print(len(consistent_edges))\n",
    "# print(len(merged_edges))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sparsified graph\n",
    "sparse_edge_colors = color_edge_section(geneGraphSparse)\n",
    "sparse_edge_styles = edge_style(geneGraphSparse)\n",
    "sparse_node_colors = nodes_apply_colormap(geneGraphSparse)\n",
    "\n",
    "pos = nx.spring_layout(geneGraphSparse, k=0.7, iterations=15, seed = 400) #scale\n",
    "fig, ax = plt.subplots(figsize=(12, 8))\n",
    "nx.draw(geneGraphSparse, pos, with_labels=False, width = 0.1, node_color=sparse_node_colors, edge_color = sparse_edge_colors, style = sparse_edge_styles, node_size=8) #, font_size = 10, font_color='black')\n",
    "\n",
    "# Showing the color bars\n",
    "for category, nodes in all_nodes.items():\n",
    "    cmap_name = base_colors[category]\n",
    "    values = list(nodes.values())\n",
    "    norm = mcolors.Normalize(vmin=min(values), vmax=max(values))\n",
    "    cmap = cm.get_cmap(cmap_name)\n",
    "    sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "    sm.set_array([])\n",
    "    plt.colorbar(sm, ax=ax, label=category)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sparsification - pruning nodes based on degree distribution\n",
    "\n",
    "# Finding all the nodes with a degree higher than 10 -- these are the only nodes that will be added to the graph\n",
    "new_nodes = []\n",
    "for node, degree in node_degree_dict.items():\n",
    "    if degree <= 100:\n",
    "        new_nodes.append(node)\n",
    "    else:\n",
    "        pass\n",
    "# Finding respective edges based on these nodes\n",
    "new_edge_list = []\n",
    "for node in new_nodes:\n",
    "    for edge in edge_list:\n",
    "        a,b = edge\n",
    "        if node == a or node == b:\n",
    "            new_edge_list.append(edge)\n",
    "\n",
    "# Initializing sparse graph and adding edges\n",
    "geneGraphSparse2 = nx.Graph()\n",
    "geneGraphSparse2.add_edges_from(new_edge_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checking length of attributes\n",
    "sparse_edge_colors2 = color_edge_section(geneGraphSparse2)\n",
    "sparse_edge_styles2 = edge_style(geneGraphSparse2)\n",
    "sparse_node_colors2 = nodes_apply_colormap(geneGraphSparse2)\n",
    "\n",
    "print(len(sparse_node_colors2))\n",
    "print(len(sparse_edge_colors2))\n",
    "print(len(sparse_edge_styles2))\n",
    "\n",
    "print(len(geneGraphSparse2.edges))\n",
    "print(len(geneGraphSparse2.nodes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Showing sparsified graph\n",
    "pos = nx.spring_layout(geneGraphSparse2, k=0.7, iterations=15, seed = 400) #scale\n",
    "fig, ax = plt.subplots(figsize=(12, 8))\n",
    "nx.draw(geneGraphSparse2, pos, with_labels=False, width = 0.1, node_color=sparse_node_colors2, edge_color = sparse_edge_colors2, style = sparse_edge_styles2, node_size=8) #, font_size = 10, font_color='black')\n",
    "\n",
    "# # Showing the color bars\n",
    "# for category, nodes in all_nodes.items():\n",
    "#     cmap_name = base_colors[category]\n",
    "#     values = list(nodes.values())\n",
    "#     norm = mcolors.Normalize(vmin=min(values), vmax=max(values))\n",
    "#     cmap = cm.get_cmap(cmap_name)\n",
    "#     sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "#     sm.set_array([])\n",
    "#     plt.colorbar(sm, ax=ax, label=category)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sparsification based on DIOPT scores\n",
    "\n",
    "# Removing all DIOPT scores above 3\n",
    "new_edges = []\n",
    "for index, edge in enumerate(first_edges):\n",
    "    diopt = edge_df.loc[index]['interaction_count']\n",
    "    if diopt >= 3:\n",
    "        new_edges.append(edge)\n",
    "\n",
    "new_edges_list= list(map(edge_tuples, new_edges))\n",
    "\n",
    "geneGraphSparse3 = nx.Graph()\n",
    "geneGraphSparse3.add_edges_from(new_edges_list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Checking length of attributes\n",
    "sparse_edge_colors3 = color_edge_section(geneGraphSparse3)\n",
    "sparse_edge_styles3 = edge_style(geneGraphSparse3)\n",
    "sparse_node_colors3 = nodes_apply_colormap(geneGraphSparse3)\n",
    "\n",
    "print(len(sparse_node_colors3))\n",
    "print(len(sparse_edge_colors3))\n",
    "print(len(sparse_edge_styles3))\n",
    "\n",
    "print(len(geneGraphSparse3.edges))\n",
    "print(len(geneGraphSparse3.nodes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Showing sparsified graph\n",
    "pos = nx.spring_layout(geneGraphSparse3, k=0.7, iterations=15, seed = 400) #scale\n",
    "fig, ax = plt.subplots(figsize=(12, 8))\n",
    "nx.draw(geneGraphSparse3, pos, with_labels=False, width = 0.1, node_color=sparse_node_colors3, edge_color = sparse_edge_colors3, style = sparse_edge_styles3, node_size=8) #, font_size = 10, font_color='black')\n",
    "\n",
    "# # Showing the color bars\n",
    "# for category, nodes in all_nodes.items():\n",
    "#     cmap_name = base_colors[category]\n",
    "#     values = list(nodes.values())\n",
    "#     norm = mcolors.Normalize(vmin=min(values), vmax=max(values))\n",
    "#     cmap = cm.get_cmap(cmap_name)\n",
    "#     sm = plt.cm.ScalarMappable(cmap=cmap, norm=norm)\n",
    "#     sm.set_array([])\n",
    "#     plt.colorbar(sm, ax=ax, label=category)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Random walk (first order)\n",
    "def random_walk(G, start_node, num_steps):\n",
    "    walk = [start_node]\n",
    "    current_node = start_node\n",
    "\n",
    "    for _ in range(num_steps):\n",
    "        neighbors = list(G.neighbors(current_node))\n",
    "        if neighbors == 0:\n",
    "            break\n",
    "        next_node = random.choice(neighbors)\n",
    "        walk.append(next_node)\n",
    "        current_node = next_node\n",
    "\n",
    "    return walk\n",
    "\n",
    "# Random walk (second order)\n",
    "def second_order_random_walk(graph, start_node, num_steps, p, q):\n",
    "    walk = [start_node]\n",
    "    current_node = start_node\n",
    "    previous_node = None\n",
    "\n",
    "    for _ in range(num_steps):\n",
    "        neighbors = list(graph.neighbors(current_node))\n",
    "        if not neighbors:\n",
    "            break\n",
    "\n",
    "        if previous_node is None:\n",
    "            # First step, no previous node\n",
    "            next_node = random.choice(neighbors)\n",
    "        else:\n",
    "            probabilities = []\n",
    "            for neighbor in neighbors:\n",
    "                if neighbor == previous_node:\n",
    "                    probabilities.append(1 / p)\n",
    "                elif graph.has_edge(previous_node, neighbor):\n",
    "                    probabilities.append(1)\n",
    "                else:\n",
    "                    probabilities.append(1 / q)\n",
    "\n",
    "            # Normalize probabilities\n",
    "            probabilities = np.array(probabilities)\n",
    "            probabilities /= probabilities.sum()\n",
    "\n",
    "            # Choose next node based on the transition probabilities\n",
    "            next_node = np.random.choice(neighbors, p=probabilities)\n",
    "\n",
    "        walk.append(next_node)\n",
    "        previous_node = current_node\n",
    "        current_node = next_node\n",
    "\n",
    "    return walk\n",
    "\n",
    "random_walk(geneGraph, 9722, 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using Node2Vec and UMAP to make nonlinear plots\n",
    "\n",
    "# Initialize node2vec model\n",
    "node2vec = Node2Vec(geneGraph, dimensions=64, walk_length=30, num_walks=200, workers=4)\n",
    "\n",
    "# # Build the vocabulary\n",
    "# model = node2vec.fit(window=10, min_count=1, batch_words=4)\n",
    "\n",
    "# # Save embeddings for later use\n",
    "# model.wv.save_word2vec_format('gene_node_embeddings.emb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ADJACENCY MATRIX\n",
    "\n",
    "verticies = len(list(geneGraph.nodes()))\n",
    "node_name_list = list(geneGraph.nodes())\n",
    "\n",
    "# (Converting the elements from node_name_list from int to str)\n",
    "node_list = []\n",
    "for node in node_name_list:\n",
    "    node = str(node)\n",
    "    node_list.append(node)\n",
    "\n",
    "# Mapping each node to an index and index to a node\n",
    "node_to_index = {node: index for index,node in enumerate(nodes_list)}\n",
    "index_to_node = {index: node for index,node in enumerate(nodes_list)}\n",
    "\n",
    "\n",
    "# Converting each node from its name to its index\n",
    "edge_index_list = []\n",
    "for node1, node2 in edge_list:\n",
    "    node1_index = node_to_index[node1]\n",
    "    node2_index = node_to_index[node2]\n",
    "    edge_index_list.append((node1_index, node2_index))\n",
    "\n",
    "# Making an empty matrix\n",
    "adj_matrix = [[0] * verticies for __ in range(verticies)]\n",
    "\n",
    "# Add values to the matrix\n",
    "for node1, node2 in edge_index_list:\n",
    "    adj_matrix[node1][node2] = 1       # Becuase this is a directed graph\n",
    "\n",
    "# Printing labels for the matrix\n",
    "print(\"\\nMatrix with Labels:\")\n",
    "print(\" \" + \" \".join(node_list))\n",
    "print(\"  \" + \"-\" * (4 * verticies - 1))\n",
    "\n",
    "# Iterate through each row of the matrix\n",
    "for i, row in enumerate(adj_matrix):\n",
    "    # Print the label for the current row node\n",
    "    print(f\"{node_name_list[i]} | \" + \"   \".join(map(str, row)))\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  },
  "vscode": {
   "interpreter": {
    "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
